src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,712 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <config.h>
+#include <json-glib/json-glib.h>
+#include <flatpak.h>
+#include <glib.h>
+#include <gnome-software.h>
+#include <string.h>
+
+#include "gs-flatpak.h"
+
+#define EXTERNAL_ASSETS_SPEC_VERSION 1
+#define JSON_SPEC_KEY "spec"
+#define JSON_RUNTIME_KEY "runtime"
+#define JSON_RUNTIME_NAME_KEY "name"
+#define JSON_RUNTIME_URL_KEY "url"
+
+#define METADATA_URL "GnomeSoftware::external-app::url"
+#define METADATA_HEADLESS_APP "GnomeSoftware::external-app::headless-app"
+#define METADATA_BUILD_DIR "GnomeSoftware::external-app::build-dir"
+#define METADATA_EXTERNAL_ASSETS "flatpack-3rdparty::external-assets"
+
+struct GsPluginData {
+	GsFlatpak	*flatpak;
+};
+
+void
+gs_plugin_initialize (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_alloc_data (plugin, sizeof(GsPluginData));
+
+	priv->flatpak = gs_flatpak_new (plugin, GS_FLATPAK_SCOPE_USER);
+
+	/* Run plugin after the flatpak ones because we need them to install
+	 * the app's headless part first */
+	gs_plugin_add_rule (plugin, GS_PLUGIN_RULE_RUN_BEFORE, "flatpak-system");
+	gs_plugin_add_rule (plugin, GS_PLUGIN_RULE_RUN_BEFORE, "flatpak-user");
+}
+
+static gboolean
+gs_app_is_flatpak (GsApp *app)
+{
+	const gchar *id = gs_app_get_id (app);
+	return id != NULL &&
+		(g_str_has_prefix (id, GS_FLATPAK_USER_PREFIX ":") ||
+		 g_str_has_prefix (id, GS_FLATPAK_SYSTEM_PREFIX ":"));
+}
+
+void
+gs_plugin_adopt_app (GsPlugin *plugin, GsApp *app)
+{
+	const gchar *id = gs_app_get_id (app);
+	if (gs_app_is_flatpak (app)) {
+		const gchar *metadata;
+		metadata = gs_app_get_metadata_item (app,
+						     METADATA_EXTERNAL_ASSETS);
+
+		if (!metadata)
+			return;
+
+		g_debug ("Adopt '%s' as an external app", id);
+		gs_app_set_management_plugin (app, gs_plugin_get_name (plugin));
+	}
+}
+
+gboolean
+gs_plugin_setup (GsPlugin *plugin, GCancellable *cancellable, GError **error)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	return gs_flatpak_setup (priv->flatpak, cancellable, error);
+}
+
+static char *
+gs_plugin_download_asset (GsPlugin *plugin,
+			  GsApp *app,
+			  const char *runtime,
+			  const char *asset,
+			  GError **error)
+{
+	gsize data_len;
+	g_autofree gchar *cache_basename = NULL;
+	g_autofree gchar *cache_fn = NULL;
+	g_autofree gchar *cache_png = NULL;
+	g_autofree gchar *data = NULL;
+	g_autoptr(AsIcon) icon = NULL;
+	g_autoptr(GdkPixbuf) pb = NULL;
+
+	/* download icons from the cdn */
+	cache_basename = g_path_get_basename (asset);
+	cache_fn = gs_utils_get_cache_filename (gs_plugin_get_name (plugin),
+						cache_basename,
+						GS_UTILS_CACHE_FLAG_NONE,
+						error);
+	if (cache_fn == NULL)
+		return FALSE;
+	if (g_file_test (cache_fn, G_FILE_TEST_EXISTS)) {
+		if (!g_file_get_contents (cache_fn, &data, &data_len, error))
```

Comment:
You could just check for `G_FILE_ERROR_NOENT` error coming from `g_file_get_contents()` instead of checking if the file exists before hand